File: cpp/2.cpp
#include <iostream>

using namespace std;

// Функция, реализующая расширенный алгоритм Евклида
int extendedEuclidean(int a, int b, int &c, int &d) {
    if (b == 0) {
        c = 1;
        d = 0;
        return a;
    }

    int s1, d1;
    int gcd = extendedEuclidean(b, a % b, s1, d1);

    // Обновляем c и d на основе результатов рекурсивного вызова
    c = d1;
    d = s1 - (a / b) * d1;

    return gcd;
}

// Функция для вычисления c такого, что c * d ≡ 1 (mod m)
int modInverse(int d, int m) {
    int c, temp;
    int gcd = extendedEuclidean(d, m, c, temp);

    if (gcd != 1) {
        // Обратный элемент не существует, если НОД(d, m) != 1
        return -1; // или выберите другой подход к обработке ошибки
    } else {
        // Найденное c будет обратным к d по модулю m
        return (c % m + m) % m;
    }
}

int main() {
    int d, m;
    cout << "Введите число d: ";
    cin >> d;
    cout << "Введите модуль m: ";
    cin >> m;

    int inverse = modInverse(d, m);

    if (inverse == -1) {
        cout << "Обратный элемент для " << d << " по модулю " << m << " не существует." << endl;
    } else {
        cout << "Обратный элемент для " << d << " по модулю " << m << " равен " << inverse << endl;
    }

    return 0;
}

File: go/2.go
package main

import "fmt"

// extendedEuclidean implements the extended Euclidean algorithm
func extendedEuclidean(a, b int) (int, int, int) {
	if b == 0 {
		return a, 1, 0
	}

	gcd, s1, d1 := extendedEuclidean(b, a%b)

	// Update c and d based on the results of the recursive call
	c := d1
	d := s1 - (a/b)*d1

	return gcd, c, d
}

// modInverse calculates c such that c*d ≡ 1 (mod m)
func modInverse(d, m int) int {
	gcd, c, _ := extendedEuclidean(d, m)

	if gcd != 1 {
		// Inverse element does not exist if gcd(d, m) != 1
		return -1 // or choose another approach to handle the error
	} else {
		// The found c will be the inverse of d modulo m
		return (c%m + m) % m
	}
}

func main() {
	var d, m int
	fmt.Print("Enter the number d: ")
	fmt.Scanln(&d)
	fmt.Print("Enter the modulus m: ")
	fmt.Scanln(&m)

	inverse := modInverse(d, m)

	if inverse == -1 {
		fmt.Printf("Inverse element for %d modulo %d does not exist.\n", d, m)
	} else {
		fmt.Printf("Inverse element for %d modulo %d is %d\n", d, m, inverse)
	}
}
